core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -0,0 +1,239 @@
+      // check the magic value
+      if (!logBuffer.hasMatchingShallowMagic(messageFormatVersion))
+        convertAndAssignOffsetsNonCompressed(logBuffer, offsetCounter, compactedTopic, now, messageTimestampType,
+          messageTimestampDiffMaxMs, messageFormatVersion)
+      else
+        // Do in-place validation, offset assignment and maybe set timestamp
+        assignOffsetsNonCompressed(logBuffer, offsetCounter, now, compactedTopic, messageTimestampType,
+          messageTimestampDiffMaxMs)
+    } else {
+      // Deal with compressed messages
+      // We cannot do in place assignment in one of the following situations:
+      // 1. Source and target compression codec are different
+      // 2. When magic value to use is 0 because offsets need to be overwritten
+      // 3. When magic value to use is above 0, but some fields of inner messages need to be overwritten.
+      // 4. Message format conversion is needed.
+
+      // No in place assignment situation 1 and 2
+      var inPlaceAssignment = sourceCodec == targetCodec && messageFormatVersion > Record.MAGIC_VALUE_V0
+
+      var maxTimestamp = Record.NO_TIMESTAMP
+      val expectedInnerOffset = new LongRef(0)
+      val validatedRecords = new mutable.ArrayBuffer[Record]
+
+      logBuffer.deepEntries(true).asScala.foreach { logEntry =>
+        val record = logEntry.record
+        validateKey(record, compactedTopic)
+
+        if (record.magic > Record.MAGIC_VALUE_V0 && messageFormatVersion > Record.MAGIC_VALUE_V0) {
+          // No in place assignment situation 3
+          // Validate the timestamp
+          validateTimestamp(record, now, messageTimestampType, messageTimestampDiffMaxMs)
+          // Check if we need to overwrite offset
+          if (logEntry.offset != expectedInnerOffset.getAndIncrement())
+            inPlaceAssignment = false
+          if (record.timestamp > maxTimestamp)
+            maxTimestamp = record.timestamp
+        }
+
+        if (sourceCodec != NoCompressionCodec && logEntry.isCompressed)
+          throw new InvalidMessageException("Compressed outer record should not have an inner record with a " +
+            s"compression attribute set: $record")
+
+        // No in place assignment situation 4
+        if (record.magic != messageFormatVersion)
+          inPlaceAssignment = false
+
+        validatedRecords += record.convert(messageFormatVersion)
+      }
+
+      if (!inPlaceAssignment) {
+        val entries = validatedRecords.map(record => LogEntry.create(offsetCounter.getAndIncrement(), record))
+        val builder = MemoryLogBuffer.builderWithEntries(messageTimestampType, CompressionType.forId(targetCodec.codec),
+          now, entries.asJava)
+        builder.close()
+        val info = builder.info
+
+        ValidationAndOffsetAssignResult(
+          validatedEntries = builder.build(),
+          maxTimestamp = info.maxTimestamp,
+          offsetOfMaxTimestamp = info.offsetOfMaxTimestamp,
```

Comment:
info.offsetOfMaxTimestamp returns the deep offset of the message with max timestamp. To be consistent with the other code path, if compression is enabled, it seems that we want to return the shallow offset? It maybe clearer to also rename ValidationAndOffsetAssignResult.offsetOfMaxTimestamp to sth like shallowOffsetOfMaxTimestamp.